src/versionbits.cpp

```diff
@@ -128,19 +128,48 @@ class VersionBitsConditionChecker : public AbstractThresholdConditionChecker {
 
 }
 
-ThresholdState VersionBitsState(const CBlockIndex* pindexPrev, const Consensus::Params& params, Consensus::DeploymentPos pos, VersionBitsCache& cache)
-{
-    return VersionBitsConditionChecker(pos).GetStateFor(pindexPrev, params, cache.caches[pos]);
-}
-
 uint32_t VersionBitsMask(const Consensus::Params& params, Consensus::DeploymentPos pos)
 {
     return VersionBitsConditionChecker(pos).Mask(params);
 }
 
 void VersionBitsCache::Clear()
 {
+    LOCK(cs);
     for (unsigned int d = 0; d < Consensus::MAX_VERSION_BITS_DEPLOYMENTS; d++) {
         caches[d].clear();
     }
 }
+
+ThresholdState VersionBitsCache::VersionBitsState(const CBlockIndex* pindexPrev, const Consensus::Params& params, Consensus::DeploymentPos pos)
+{
+    LOCK(cs);
+    return VersionBitsConditionChecker(pos).CachedGetStateFor(pindexPrev, params, caches[pos]);
+}
+
+std::vector<ThresholdState> VersionBitsCache::GetAllThresholdStates(const CBlockIndex* pindexPrev, const Consensus::Params& params)
+{
+    std::vector<ThresholdState> ret;
+    ret.reserve((int)Consensus::MAX_VERSION_BITS_DEPLOYMENTS);
+    LOCK(cs);
+    for (int i = 0; i < (int)Consensus::MAX_VERSION_BITS_DEPLOYMENTS; ++i) {
```

Comment:
Aligning indices everywhere is a bit clunky, and will get worse with more deployments. Just return a std::map<DeploymentPos,ThresholdState> ? Then the caller can iterate cleanly.